docs/resources/server/loottables/datagen.md

```diff
@@ -0,0 +1,200 @@
+- `TagEntry#tagContents`: Provides a stack of one item, per item in the tag. The tag can be specified as a parameter.
+- `TagEntry#expandTag`: Similar to `#tagContents`, however one loot entry is created for each item in the tag instead of one entry that drops all items at once.
+- `LootTableReference#lootTableReference`: Rolls the specified loot table and returns the resulting item stacks as one entry.
+- `DynamicLoot#dynamicEntry`: Tells the game to call on a dynamic drops provider by the specified name. See [Using a Loot Table][useloottable] for more info.
+
+The composites then use the singletons or other composites in the following ways:
+
+- `EntryGroup#list`: Accepts a vararg of other `LootPoolEntryContainer.Builder<?>`s. Runs all of its children, regardless of if they succeed or not.
+- `AlternativesEntry#alternatives`: Accepts a vararg of other `LootPoolEntryContainer.Builder<?>`s. Runs its children in order until one succeeds.
+- `SequentialEntry#sequential`: The opposite of `AlternativesEntry`. Accepts a vararg of other `LootPoolEntryContainer.Builder<?>`s, and runs its children in order until one fails.
+
+All `LootPoolEntryContainer.Builder<?>`s additionally allow adding loot conditions and loot functions by calling `#when` and `#apply`, respectively, similar to loot pools (see above). Additionally, be aware that mods may add custom loot entry providers.
+
+### Number Providers
+
+Number providers define how a number is obtained. This is done instead of simply using constant integers and floats in order to bring more randomness to loot tables. Vanilla defines four number providers, however more may be added by mods.
+
+- `ConstantValue#exactly`: A constant integer or float value.
+- `UniformGenerator#between`: Uniformly-distributed random integer or float values, with min and max values set. All values between min and max have the same chance to appear.
+- `BinomialDistributionGenerator#binomial`: Binomially-distributed random integer values, with n and p values set. See [Binomial Distribution][binomial] for more information on what these values mean.
+- `ScoreboardValue#fromScoreboard`: Given an entity target, a score name and (optionally) a scale value, retrieves the given scoreboard value for the entity target, multiplying it with the given scale value (if available).
+
+### `BlockLootSubProvider`
+
+`BlockLootSubProvider` is an abstract helper class containing many helpers for creating common block loot tables, e.g. single item drops (`#createSingleItemTable`), silk touch-only drops (`#createSilkTouchOnlyTable`), drops for slab-like blocks (`#createSlabItemTable`), and many more. Unfortunately, setting up a `BlockLootSubProvider` for modded usage involves more boilerplate:
+
+```java
+public class MyBlockLootSubProvider extends BlockLootSubProvider {
+    // A list of blocks to generate tables for. Must be a mutable list because we will add values to it below.
+    private final List<Block> blocks = new ArrayList<>();
```

Comment:
No, this is just bad practice.  `getKnownBlocks` should be the blocks that must have a loot table for it. You shouldn't be purposely working around the validator.